jicofo-common/src/main/java/org/jitsi/impl/protocol/xmpp/ChatRoomImpl.kt

```diff
@@ -0,0 +1,769 @@
+                logger.warn("Removing " + members.size + " stale members.")
+                members.clear()
+            }
+        }
+        synchronized(this) {
+            role = null
+            lastPresenceSent = null
+            meetingId = null
+            logger.addContext("meeting_id", "")
+            isBreakoutRoom = false
+            mainRoom = null
+            avModeration.reset()
+        }
+    }
+
+    @Throws(SmackException::class, XMPPException::class, InterruptedException::class)
+    private fun joinAs(nickname: Resourcepart) {
+        myOccupantJid = JidCreate.entityFullFrom(roomJid, nickname)
+        presenceInterceptor = org.jivesoftware.smack.util.Consumer { presenceBuilder: PresenceBuilder ->
+            // The initial presence sent by smack contains an empty "x"
+            // extension. If this extension is included in a subsequent stanza,
+            // it indicates that the client lost its synchronization and causes
+            // the MUC service to re-send the presence of each occupant in the
+            // room.
+            synchronized(this@ChatRoomImpl) {
+                val p = presenceBuilder.build()
+                p.removeExtension(
+                    MUCInitialPresence.ELEMENT,
+                    MUCInitialPresence.NAMESPACE
+                )
+                lastPresenceSent = p.asBuilder()
+            }
+        }
+        if (muc!!.isJoined) {
+            muc.leave()
+        }
+        muc.addPresenceInterceptor(presenceInterceptor)
+        muc.createOrJoin(nickname)
+        val config = muc.configurationForm
+
+        // Read breakout rooms options
+        val isBreakoutRoomField = config.getField(MucConfigFields.IS_BREAKOUT_ROOM)
+        if (isBreakoutRoomField != null) {
+            isBreakoutRoom = java.lang.Boolean.parseBoolean(isBreakoutRoomField.firstValue)
+            if (isBreakoutRoom) {
+                val mainRoomField = config.getField(MucConfigFields.MAIN_ROOM)
+                if (mainRoomField != null) {
+                    mainRoom = mainRoomField.firstValue
+                }
+            }
+        }
+
+        // Read meetingId
+        val meetingIdField = config.getField(MucConfigFields.MEETING_ID)
+        if (meetingIdField != null) {
+            meetingId = meetingIdField.firstValue
+            if (meetingId != null) {
+                logger.addContext("meeting_id", meetingId)
+            }
+        }
+
+        // Make the room non-anonymous, so that others can recognize focus JID
+        val answer = config.fillableForm
+        answer.setAnswer(MucConfigFields.WHOIS, "anyone")
+        muc.sendConfigurationForm(answer)
+    }
+
+    private fun leave(reason: String, jid: EntityBareJid) {
+        logger.info("Leave, reason: $reason alt-jid: $jid")
+        leave()
+    }
+
+    override fun leave() {
+        if (presenceInterceptor != null) {
+            muc.removePresenceInterceptor(presenceInterceptor)
+        }
```

Comment:
Should be expressable with a .?let I think (which might make it possible to let removePresenceInterceptor's argument non-nullable?)